pkg/controller/volume/pvcprotection/pvc_protection_controller.go

```diff
@@ -109,45 +173,64 @@ func (c *Controller) Run(ctx context.Context, workers int) {
 		return
 	}
 
+	go wait.UntilWithContext(ctx, c.runMainWorker, time.Second)
 	for i := 0; i < workers; i++ {
-		go wait.UntilWithContext(ctx, c.runWorker, time.Second)
+		go wait.UntilWithContext(ctx, c.runProcessNamespaceWorker, time.Second)
 	}
 
 	<-ctx.Done()
 }
 
-func (c *Controller) runWorker(ctx context.Context) {
+func (c *Controller) runMainWorker(ctx context.Context) {
 	for c.processNextWorkItem(ctx) {
 	}
 }
 
+func (c *Controller) runProcessNamespaceWorker(ctx context.Context) {
+	for c.processPVCsByNamespace(ctx) {
+	}
+}
+
 // processNextWorkItem deals with one pvcKey off the queue.  It returns false when it's time to quit.
 func (c *Controller) processNextWorkItem(ctx context.Context) bool {
-	pvcKey, quit := c.queue.Get()
-	if quit {
-		return false
+	newNamespaces := make([]string, 0)
+	queueLength := c.queue.Len()
+	for i := 0; i < queueLength; i++ {
+		pvcKey, _ := c.queue.Get()
```

Comment:
I guess I can add this check for consistency, but it seems like a no-op to me, as quit is only true when the queue length is 0 and it is shutting down (https://github.com/kubernetes/kubernetes/blob/3a8a60eba29940e26ac8db52329a91ba87305114/staging/src/k8s.io/client-go/util/workqueue/queue.go#L281), but when we're in this loop, queueLength must already be greater than 0.

Reply:
My concern in reading the documentation of workqueue Len: https://pkg.go.dev/k8s.io/client-go/util/workqueue#Type.Len

```
Len returns the current queue length, for informational purposes only. You shouldn't e.g. gate a call to Add() or Get() on Len() being a particular value, that can't be synchronized properly.
```

If Len() doesn't accurately reflect the queue length (eg: in a concurrent consumer situation) we may not quit when appropriate. I'm not sure if this actually matters since we have a single consumer, but I would prefer to stick to the API contract when possible.

> I also just added a check for whether the queue is shutting down before this loop though to catch this case.

I'm not seeing this at first glance. Can you post a permalink to where you made this change?

Reply:
Resolved with the latest commit.